src/graphapi/graphapi_tensor.cpp

```diff
@@ -0,0 +1,312 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/graphapi/graphapi_tensor.hpp>
+#include <miopen/errors.hpp>
+
+#include <algorithm>
+
+namespace miopen {
+
+namespace graphapi {
+
+TensorBuilder& TensorBuilder::setDataType(miopenDataType_t dataType) &
+{
+    mTensor.setDataType(dataType);
+    mDataTypeSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setDim(const std::vector<int64_t>& dimensions) &
+{
+    if(dimensions.empty() ||
+       std::any_of(dimensions.cbegin(), dimensions.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setDimensions(dimensions);
+    mDimensionsSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setDim(std::vector<int64_t>&& dimensions) &
+{
+    if(dimensions.empty() ||
+       std::any_of(dimensions.cbegin(), dimensions.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setDimensions(std::move(dimensions));
+    mDimensionsSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setStride(const std::vector<int64_t>& strides) &
+{
+    if(strides.empty() ||
+       std::any_of(strides.cbegin(), strides.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setStrides(strides);
+    mStridesSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setStride(std::vector<int64_t>&& strides) &
+{
+    if(strides.empty() ||
+       std::any_of(strides.cbegin(), strides.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setStrides(std::move(strides));
+    mStridesSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setId(int64_t id) &
+{
+    mTensor.setId(id);
+    mUniqueIdSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setVirtual(bool isVirtual) &
+{
+    mTensor.setVirtual(isVirtual);
+    return *this;
+}
+
+Tensor TensorBuilder::build() const&
+{
+    if(!mUniqueIdSet || !mDataTypeSet || !mDimensionsSet || !mStridesSet ||
+       mTensor.getDimensions().size() != mTensor.getStrides().size())
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    return mTensor;
+}
+
+Tensor TensorBuilder::build() &&
+{
+    if(!mUniqueIdSet || !mDataTypeSet || !mDimensionsSet || !mStridesSet ||
+       mTensor.getDimensions().size() != mTensor.getStrides().size())
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    return std::move(mTensor);
```

Comment:
Better to not use `std::move` here and let the compiler pick.

Reply:
mTensor is l-value here and won't be moved without std::move.
copy-elision eliminates additional creation; mTensor was created already and its creation cannot be eliminated.

without std::move the object to be returned will always be copy-constructed.

Reply:
seems like you're right.